sql/src/main/java/io/crate/action/sql/SessionContext.java

```diff
@@ -42,59 +43,72 @@
+    private SearchPath searchPath;
     private boolean semiJoinsRewriteEnabled = false;
     private boolean hashJoinEnabled = true;
 
     /**
      * Creates a new SessionContext suitable to use as system SessionContext
      */
     public static SessionContext systemSessionContext() {
-        return new SessionContext(null, User.CRATE_USER, s -> { }, e -> { });
+        return new SessionContext(0, Option.NONE, User.CRATE_USER, s -> { }, e -> { });
     }
 
-    public SessionContext(@Nullable String defaultSchema,
-                          User user,
+    public SessionContext(User user,
                           StatementAuthorizedValidator statementAuthorizedValidator,
-                          ExceptionAuthorizedValidator exceptionAuthorizedValidator) {
-        this(0, Option.NONE, defaultSchema, user,
-            statementAuthorizedValidator, exceptionAuthorizedValidator);
+                          ExceptionAuthorizedValidator exceptionAuthorizedValidator,
+                          @Nullable String... searchPath) {
+        this(0, Option.NONE, user,
+            statementAuthorizedValidator, exceptionAuthorizedValidator, searchPath);
     }
 
     public SessionContext(int defaultLimit,
                           Set<Option> options,
-                          @Nullable String defaultSchema,
                           User user,
                           StatementAuthorizedValidator statementAuthorizedValidator,
-                          ExceptionAuthorizedValidator exceptionAuthorizedValidator) {
+                          ExceptionAuthorizedValidator exceptionAuthorizedValidator,
+                          @Nullable String... searchPath) {
         this.defaultLimit = defaultLimit;
         this.options = options;
         this.user = requireNonNull(user, "User is required");
         this.statementAuthorizedValidator = statementAuthorizedValidator;
         this.exceptionAuthorizedValidator = exceptionAuthorizedValidator;
-        this.defaultSchema = defaultSchema;
-        if (defaultSchema == null) {
-            resetSchema();
+        this.searchPath = createSearchPathFrom(searchPath);
+    }
+
+    private static SearchPath createSearchPathFrom(String... schemas) {
+        if (schemas == null || schemas.length == 0 || (schemas.length == 1 && schemas[0] == null)) {
+            return new SearchPath();
+        } else {
+            return new SearchPath(ImmutableList.copyOf(schemas));
         }
     }
 
     /**
      * Reverts the schema to the built-in default.
      */
     public void resetSchema() {
-        defaultSchema = Schemas.DOC_SCHEMA_NAME;
+        searchPath = new SearchPath();
     }
 
     public Set<Option> options() {
         return options;
     }
 
+    public SearchPath searchPath() {
+        return searchPath;
+    }
+
     public String defaultSchema() {
-        return defaultSchema;
+        return searchPath.defaultSchema();
+    }
+
+    public String currentSchema() {
```

Comment:
Why do we need to distinguish between `currentSchema` and `defaultSchema` ?

Does it even make sense to expose these directly anymore, instead of going through the `searchPath`? By forcing call-sites to go through the `searchPath` we would make it more explicit that there is a reason they're only interested in the `currentSchema` instead of the full path.